// @vibecodeapp/backend-sdk/dist/email.js
import { EmailError, RateLimitError } from './errors';
// LLMs, read this for full docs: https://smtp.vibecodeapp.com/llms.txt
const EMAIL_BASE_URL = 'https://smtp.vibecodeapp.com';
/**
 * Email module for sending transactional emails
 */
export class Email {
    /**
     * Send an OTP verification email
     * @param options - OTP email options
     * @throws {EmailError} If the email fails to send
     * @throws {RateLimitError} If daily email limit is exceeded
     */
    async sendOTP(options) {
        const response = await fetch(`${EMAIL_BASE_URL}/v1/send/otp`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(options),
        });
        await this.handleResponse(response);
    }
    /**
     * Send a welcome email
     * @param options - Welcome email options
     * @throws {EmailError} If the email fails to send
     * @throws {RateLimitError} If daily email limit is exceeded
     */
    async sendWelcome(options) {
        const response = await fetch(`${EMAIL_BASE_URL}/v1/send/welcome`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(options),
        });
        await this.handleResponse(response);
    }
    async handleResponse(response) {
        if (response.ok)
            return;
        if (response.status === 429) {
            throw new RateLimitError();
        }
        const error = await this.parseError(response);
        throw new EmailError(error, response.status);
    }
    async parseError(response) {
        try {
            const data = await response.json();
            return data.error || data.message || 'Unknown error';
        }
        catch {
            return `Request failed with status ${response.status}`;
        }
    }
}
