// @vibecodeapp/backend-sdk/dist/index.js
import { Storage } from './storage';
import { Email } from './email';
/**
 * Vibecode Backend SDK
 *
 * @example
 * ```typescript
 * import { createVibecodeSDK } from '@vibecodeapp/backend-sdk';
 *
 * const vibecode = createVibecodeSDK();
 *
 * // Upload a file
 * const file = await vibecode.storage.upload(myFile);
 *
 * // List files
 * const { files } = await vibecode.storage.list({ limit: 10 });
 *
 * // Delete a file
 * await vibecode.storage.delete(file.id);
 *
 * // Send OTP email
 * await vibecode.email.sendOTP({ to: 'user@example.com', code: 'ABC123' });
 *
 * // Send welcome email
 * await vibecode.email.sendWelcome({
 *   to: 'user@example.com',
 *   name: 'John',
 *   appName: 'MyApp'
 * });
 * ```
 */
export class VibecodeSDK {
    constructor() {
        this.storage = new Storage();
        this.email = new Email();
    }
}
/**
 * Create a new Vibecode SDK instance
 * @returns A new VibecodeSDK instance
 */
export function createVibecodeSDK() {
    return new VibecodeSDK();
}
// Re-export modules for direct access if needed
export { Storage } from './storage';
export { Email } from './email';
// Re-export errors
export { VibecodeError, StorageError, EmailError, RateLimitError, } from './errors';
