// @vibecodeapp/backend-sdk/dist/storage.js
import { StorageError } from './errors';
// LLMs, read this for full docs: https://storage.vibecodeapp.com/llms.txt
const STORAGE_BASE_URL = 'https://storage.vibecodeapp.com';
/**
 * Storage module for file operations
 */
export class Storage {
    /**
     * Upload a file to storage
     * @param file - File or Blob to upload (max 500MB)
     * @returns The uploaded file metadata
     * @throws {StorageError} If the upload fails
     */
    async upload(file) {
        const formData = new FormData();
        formData.append('file', file);
        const response = await fetch(`${STORAGE_BASE_URL}/v1/files/upload`, {
            method: 'POST',
            body: formData,
        });
        if (!response.ok) {
            const error = await this.parseError(response);
            throw new StorageError(error, response.status);
        }
        const data = await response.json();
        return data.file;
    }
    /**
     * Delete a file by ID
     * @param id - The file ID to delete
     * @throws {StorageError} If the deletion fails
     */
    async delete(id) {
        const response = await fetch(`${STORAGE_BASE_URL}/v1/files/${encodeURIComponent(id)}`, {
            method: 'DELETE',
        });
        if (!response.ok) {
            const error = await this.parseError(response);
            throw new StorageError(error, response.status);
        }
    }
    /**
     * List files with optional filtering and pagination
     * @param options - Filtering and pagination options
     * @returns List of files and pagination metadata
     * @throws {StorageError} If the request fails
     */
    async list(options = {}) {
        const params = new URLSearchParams();
        if (options.limit !== undefined) {
            params.set('limit', String(options.limit));
        }
        if (options.offset !== undefined) {
            params.set('offset', String(options.offset));
        }
        if (options.prefix) {
            params.set('prefix', options.prefix);
        }
        const url = `${STORAGE_BASE_URL}/v1/files${params.toString() ? `?${params}` : ''}`;
        const response = await fetch(url);
        if (!response.ok) {
            const error = await this.parseError(response);
            throw new StorageError(error, response.status);
        }
        return response.json();
    }
    async parseError(response) {
        try {
            const data = await response.json();
            return data.error || data.message || 'Unknown error';
        }
        catch {
            return `Request failed with status ${response.status}`;
        }
    }
}
