// @vibecodeapp/proxy/dist/manual.js
/**
 * Manual initialization entry point for the Vibecode proxy.
 * Use this when you need full control over when the proxy initializes.
 *
 * Usage:
 *   import { init } from '@vibecodeapp/proxy/manual'
 *   init({
 *     projectId: 'your-project-id',
 *     excludeDomains: ['api.openai.com'], // Use your own OpenAI key
 *   })
 */
const PROXY_HOST = "proxy.vibecodeapp.com";
const PROXY_DOMAINS_URL = "https://proxy.vibecodeapp.com/v1/domains";
const DEFAULT_PROXIED_DOMAINS = [
    "api.openai.com",
    "api.anthropic.com",
    "api.x.ai",
    "amazonaws.com",
    "generativelanguage.googleapis.com",
    "api.elevenlabs.io",
];
// State
let proxiedDomains = new Set(DEFAULT_PROXIED_DOMAINS);
let excludedDomains = new Set();
let proxyProjectId = null;
let isInitialized = false;
// Store the original fetch before any patching
const originalFetch = globalThis.fetch;
/**
 * Check if a host is in a domain set (exact match or subdomain)
 */
function matchesDomain(host, domains) {
    for (const domain of domains) {
        if (host === domain || host.endsWith(`.${domain}`)) {
            return true;
        }
    }
    return false;
}
/**
 * Check if a host should be proxied
 */
function shouldProxy(host) {
    if (matchesDomain(host, excludedDomains)) {
        return false;
    }
    return matchesDomain(host, proxiedDomains);
}
/**
 * Transform a URL to go through the proxy
 */
function transformUrl(originalUrl) {
    const newHost = `${originalUrl.host}.${PROXY_HOST}`;
    const transformed = new URL(originalUrl.toString());
    transformed.host = newHost;
    if (proxyProjectId) {
        transformed.username = proxyProjectId;
        transformed.password = proxyProjectId;
    }
    return transformed;
}
/**
 * Patched fetch that routes matching requests through the proxy
 */
async function patchedFetch(input, init) {
    let url;
    let modifiedInit = init ? { ...init } : {};
    if (typeof input === "string") {
        try {
            url = new URL(input);
        }
        catch {
            return originalFetch(input, init);
        }
    }
    else if (input instanceof URL) {
        url = input;
    }
    else if (input instanceof Request) {
        url = new URL(input.url);
        modifiedInit = {
            method: input.method,
            headers: input.headers,
            body: input.body,
            ...modifiedInit,
        };
    }
    else {
        return originalFetch(input, init);
    }
    if (shouldProxy(url.host)) {
        const transformedUrl = transformUrl(url);
        if (proxyProjectId) {
            const headers = new Headers(modifiedInit.headers);
            headers.set("X-Vibecode-Project", proxyProjectId);
            modifiedInit.headers = headers;
        }
        console.log(`[Vibecode Proxy] ${url.host} → ${transformedUrl.host}`);
        return originalFetch(transformedUrl.toString(), modifiedInit);
    }
    return originalFetch(input, init);
}
/**
 * Fetch the list of proxied domains from the server
 */
async function fetchProxiedDomains() {
    try {
        const response = await originalFetch(PROXY_DOMAINS_URL);
        if (response.ok) {
            const data = await response.json();
            let domains = null;
            if (Array.isArray(data)) {
                domains = data;
            }
            else if (typeof data === "object" &&
                data !== null &&
                "domains" in data &&
                Array.isArray(data.domains)) {
                domains = data.domains;
            }
            if (domains) {
                proxiedDomains = new Set(domains);
                console.log(`[Vibecode Proxy] Loaded ${proxiedDomains.size} proxied domains`);
            }
        }
    }
    catch {
        // Silently use defaults
    }
}
/**
 * Patch undici's fetch for Node.js compatibility
 */
async function patchUndici() {
    try {
        // @ts-ignore - undici may not be installed
        const undici = (await import("undici"));
        if (undici && typeof undici.fetch === "function") {
            const originalUndici = undici.fetch;
            undici.fetch = async function (input, init) {
                let url;
                let modifiedInit = init
                    ? { ...init }
                    : {};
                if (typeof input === "string") {
                    try {
                        url = new URL(input);
                    }
                    catch {
                        return originalUndici(input, init);
                    }
                }
                else if (input instanceof URL) {
                    url = input;
                }
                else if (input instanceof Request) {
                    url = new URL(input.url);
                    modifiedInit = {
                        method: input.method,
                        headers: input.headers,
                        body: input.body,
                        ...modifiedInit,
                    };
                }
                else {
                    return originalUndici(input, init);
                }
                if (shouldProxy(url.host)) {
                    const transformedUrl = transformUrl(url);
                    if (proxyProjectId) {
                        const headers = new Headers(modifiedInit.headers);
                        headers.set("X-Vibecode-Project", proxyProjectId);
                        modifiedInit.headers = headers;
                    }
                    console.log(`[Vibecode Proxy] ${url.host} → ${transformedUrl.host}`);
                    return originalUndici(transformedUrl.toString(), modifiedInit);
                }
                return originalUndici(input, init);
            };
            console.log("[Vibecode Proxy] Patched undici");
        }
    }
    catch {
        // undici not available
    }
}
// Read credentials from environment
function getEnv(key) {
    if (typeof process !== "undefined" && process.env) {
        return process.env[key];
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const g = globalThis;
    if (g.Deno?.env) {
        return g.Deno.env.get(key);
    }
    return undefined;
}
/**
 * Initialize the proxy with custom settings.
 * Unlike the default import, this does NOT auto-patch fetch.
 *
 * @example
 * import { init } from '@vibecodeapp/proxy/manual'
 * init({
 *   projectId: 'your-project-id',
 *   excludeDomains: ['api.openai.com'],
 * })
 */
export function init(config = {}) {
    if (isInitialized) {
        console.warn("[Vibecode Proxy] Already initialized, reconfiguring...");
    }
    // Set project ID
    proxyProjectId =
        config.projectId ??
            getEnv("VIBECODE_PROJECT_ID") ??
            getEnv("EXPO_PUBLIC_VIBECODE_PROJECT_ID") ??
            getEnv("VITE_VIBECODE_PROJECT_ID") ??
            null;
    // Set excluded domains
    if (config.excludeDomains) {
        excludedDomains = new Set(config.excludeDomains);
    }
    // Set proxied domains
    if (config.proxiedDomains) {
        proxiedDomains = new Set(config.proxiedDomains);
    }
    // Patch global fetch
    globalThis.fetch = patchedFetch;
    if (proxyProjectId) {
        console.log(`[Vibecode Proxy] Initialized for project: ${proxyProjectId}`);
    }
    else {
        console.log("[Vibecode Proxy] Initialized (no project ID set)");
    }
    if (excludedDomains.size > 0) {
        console.log(`[Vibecode Proxy] Excluding domains: ${[...excludedDomains].join(", ")}`);
    }
    // Background tasks
    if (!config.skipDomainFetch) {
        fetchProxiedDomains();
    }
    if (!config.skipUndici) {
        patchUndici();
    }
    isInitialized = true;
}
/**
 * Get the original unpatched fetch
 */
export function getOriginalFetch() {
    return originalFetch;
}
/**
 * Check if the proxy has been initialized
 */
export function isProxyInitialized() {
    return isInitialized;
}
