// @vibecodeapp/proxy/dist/proxy.js
/**
 * Transparent proxy for all outgoing fetch requests.
 * Auto-patches global fetch on import - just import and go.
 *
 * Usage:
 *   import '@vibecodeapp/proxy'
 *
 * For more control:
 *   import { configure } from '@vibecodeapp/proxy'
 *   configure({ excludeDomains: ['api.openai.com'] })
 *
 * For manual initialization (no auto-patch):
 *   import { init } from '@vibecodeapp/proxy/manual'
 *
 * Set credentials via environment variables:
 *   VIBECODE_PROJECT_ID - Your project ID (used for both auth username and password)
 */
// Read environment variables (works in Node, Bun, Deno)
function getEnv(key) {
    // Node.js / Bun
    if (typeof process !== "undefined" && process.env) {
        return process.env[key];
    }
    // Deno
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const g = globalThis;
    if (g.Deno?.env) {
        return g.Deno.env.get(key);
    }
    return undefined;
}
const PROXY_HOST = getEnv("NODE_ENV") === "production"
    ? "cloudproxy.vibecodeapp.com"
    : "proxy.vibecodeapp.com";
const PROXY_DOMAINS_URL = `https://${PROXY_HOST}/v1/domains`;
const DEFAULT_PROXIED_DOMAINS = [
    "api.openai.com",
    "api.anthropic.com",
    "api.x.ai",
    "amazonaws.com",
    "generativelanguage.googleapis.com",
    "api.elevenlabs.io",
];
// State
let proxiedDomains = new Set(DEFAULT_PROXIED_DOMAINS);
let excludedDomains = new Set();
let proxyProjectId = null;
// Store the original fetch before patching
const originalFetch = globalThis.fetch;
/**
 * Check if a host is in a domain set (exact match or subdomain)
 */
function matchesDomain(host, domains) {
    for (const domain of domains) {
        if (host === domain || host.endsWith(`.${domain}`)) {
            return true;
        }
    }
    return false;
}
/**
 * Check if a host should be proxied
 */
function shouldProxy(host) {
    // Check exclusions first
    if (matchesDomain(host, excludedDomains)) {
        return false;
    }
    return matchesDomain(host, proxiedDomains);
}
/**
 * Transform a URL to go through the proxy
 */
function transformUrl(originalUrl) {
    const newHost = `${originalUrl.host}.${PROXY_HOST}`;
    const transformed = new URL(originalUrl.toString());
    transformed.host = newHost;
    // Use project ID for both username and password
    if (proxyProjectId) {
        transformed.username = proxyProjectId;
        transformed.password = proxyProjectId;
    }
    return transformed;
}
/**
 * Patched fetch that routes matching requests through the proxy
 */
async function patchedFetch(input, init) {
    let url;
    let modifiedInit = init ? { ...init } : {};
    // Parse the URL from various input types
    if (typeof input === "string") {
        try {
            url = new URL(input);
        }
        catch {
            // Relative URL or invalid - pass through
            return originalFetch(input, init);
        }
    }
    else if (input instanceof URL) {
        url = input;
    }
    else if (input instanceof Request) {
        url = new URL(input.url);
        // Merge request properties with init
        modifiedInit = {
            method: input.method,
            headers: input.headers,
            body: input.body,
            ...modifiedInit,
        };
    }
    else {
        // Fallback to original fetch for unknown types
        return originalFetch(input, init);
    }
    // Check if this request should be proxied
    if (shouldProxy(url.host)) {
        const transformedUrl = transformUrl(url);
        // Add X-Vibecode-Project header if we have a project ID
        if (proxyProjectId) {
            const headers = new Headers(modifiedInit.headers);
            headers.set("X-Vibecode-Project", proxyProjectId);
            modifiedInit.headers = headers;
        }
        console.log(`[Vibecode Proxy] ${url.host} → ${transformedUrl.host}`);
        return originalFetch(transformedUrl.toString(), modifiedInit);
    }
    // Not proxied, use original fetch
    return originalFetch(input, init);
}
/**
 * Fetch the list of proxied domains from the server (runs in background)
 */
async function fetchProxiedDomains() {
    try {
        const response = await originalFetch(PROXY_DOMAINS_URL);
        if (response.ok) {
            const data = await response.json();
            let domains = null;
            if (Array.isArray(data)) {
                domains = data;
            }
            else if (typeof data === "object" &&
                data !== null &&
                "domains" in data &&
                Array.isArray(data.domains)) {
                domains = data.domains;
            }
            if (domains) {
                proxiedDomains = new Set(domains);
                console.log(`[Vibecode Proxy] Loaded ${proxiedDomains.size} proxied domains`);
            }
        }
    }
    catch {
        // Silently use defaults - don't spam console
    }
}
/**
 * Patch undici's fetch for Node.js compatibility
 * OpenAI SDK and others may use undici directly
 */
async function patchUndici() {
    try {
        // Dynamic import to avoid errors in Bun/browser
        // @ts-ignore - undici may not be installed
        const undici = await import("undici");
        if (undici && typeof undici.fetch === "function") {
            const originalUndici = undici.fetch;
            undici.fetch = async function (input, init) {
                let url;
                let modifiedInit = init ? { ...init } : {};
                if (typeof input === "string") {
                    try {
                        url = new URL(input);
                    }
                    catch {
                        return originalUndici(input, init);
                    }
                }
                else if (input instanceof URL) {
                    url = input;
                }
                else if (input instanceof Request) {
                    url = new URL(input.url);
                    modifiedInit = {
                        method: input.method,
                        headers: input.headers,
                        body: input.body,
                        ...modifiedInit,
                    };
                }
                else {
                    return originalUndici(input, init);
                }
                if (shouldProxy(url.host)) {
                    const transformedUrl = transformUrl(url);
                    if (proxyProjectId) {
                        const headers = new Headers(modifiedInit.headers);
                        headers.set("X-Vibecode-Project", proxyProjectId);
                        modifiedInit.headers = headers;
                    }
                    console.log(`[Vibecode Proxy] ${url.host} → ${transformedUrl.host}`);
                    return originalUndici(transformedUrl.toString(), modifiedInit);
                }
                return originalUndici(input, init);
            };
            console.log("[Vibecode Proxy] Patched undici");
        }
    }
    catch {
        // undici not available - that's fine
    }
}
/**
 * Set proxy project ID programmatically
 */
export function setProxyProjectId(projectId) {
    proxyProjectId = projectId;
}
/**
 * Set proxied domains programmatically
 */
export function setProxiedDomains(domains) {
    proxiedDomains = new Set(domains);
}
/**
 * Set excluded domains programmatically
 */
export function setExcludedDomains(domains) {
    excludedDomains = new Set(domains);
}
/**
 * Add domains to the exclusion list
 */
export function addExcludedDomains(domains) {
    for (const domain of domains) {
        excludedDomains.add(domain);
    }
}
/**
 * Get the original unpatched fetch
 */
export function getOriginalFetch() {
    return originalFetch;
}
/**
 * Configure the proxy with custom settings.
 * Call this after importing to customize behavior.
 *
 * @example
 * import { configure } from '@vibecodeapp/proxy'
 * configure({
 *   excludeDomains: ['api.openai.com'], // Use your own OpenAI key
 * })
 */
export function configure(config) {
    if (config.projectId !== undefined) {
        proxyProjectId = config.projectId;
    }
    if (config.excludeDomains !== undefined) {
        excludedDomains = new Set(config.excludeDomains);
    }
    if (config.proxiedDomains !== undefined) {
        proxiedDomains = new Set(config.proxiedDomains);
    }
}
// === Auto-initialize on import ===
// Set project ID from environment
proxyProjectId = getEnv("VIBECODE_PROJECT_ID") ?? getEnv("EXPO_PUBLIC_VIBECODE_PROJECT_ID") ?? getEnv("VITE_VIBECODE_PROJECT_ID") ?? null;
// Patch global fetch immediately (synchronous)
globalThis.fetch = patchedFetch;
// Log initialization
if (proxyProjectId) {
    console.log(`[Vibecode Proxy] Initialized for project: ${proxyProjectId} and proxy host: ${PROXY_HOST}`);
}
else {
    console.log("[Vibecode Proxy] Initialized (no project ID set) and proxy host: ${PROXY_HOST}");
}
// Background tasks (non-blocking)
fetchProxiedDomains();
patchUndici();
